### PR TITLE
Bump ForwardDiff to 0.3.x

### PR DESCRIPTION
**Pull request by jrevels**
This is needed for https://github.com/jeff-regier/Celeste.jl/pull/445.

### CODE REVIEW

**Comment by mlubin** [comment]
https://github.com/JuliaOpt/Optim.jl/issues/310

File: src/optimize.jl [hunk 1]
```diff
@@ -145,17 +145,19 @@ function optimize{T, M <: Union{FirstOrderSolver, SecondOrderSolver}}(f::Functio
             error("No gradient or Hessian was provided. Either provide a gradient and Hessian, set autodiff = true in the OptimizationOptions if applicable, or choose a solver that doesn't require a Hessian.")
         end
     else
-        g!(x, out) = ForwardDiff.gradient!(out, f, x)
+        cfg = ForwardDiff.GradientConfig(initial_x)
```
Comment (jrevels) [inline src/optimize.jl]: This line (and similar lines that I added) assumes that `initial_x` and any future `x` will all have the same type/shape. Is this a valid assumption (just wanted to make sure)?

File: src/optimize.jl [hunk 2]
```diff
@@ -145,17 +145,19 @@ function optimize{T, M <: Union{FirstOrderSolver, SecondOrderSolver}}(f::Functio
...
+        g! = (x, out) -> ForwardDiff.gradient!(out, f, x, cfg)
 
-        function fg!(x, out)
-            gr_res = ForwardDiff.GradientResult(zero(T),out)
-            ForwardDiff.gradient!(gr_res, f, x)
+        fg! = (x, out) -> begin
+            gr_res = ForwardDiff.GradientResult(zero(T), out)
```
Comment (KristofferC) [inline src/optimize.jl]: Can't this be created outside `fg!` to not recreate it in every call?

**Comment by pkofod** [comment]
Hmm, the interface seems to be more complicated now than before. It seems odd that such a large change was included in a patch tag. I havn't had time to look at the changes from 3.0 to 3.1 though, so cannot comment any further right now.

**Comment by jrevels** [comment]
> It seems odd that such a large change was included in a patch tag. I havn't had time to look at the changes from 3.0 to 3.1 though, so cannot comment any further right now.

This is updating from ForwardDiff v0.2 to ForwardDiff v0.3 - I'm not going to start bumping major versions of packages I maintain until Julia itself is v1.0 (seems to be the same strategy used in Optim).

The API didn't change too much. In fact, the only actual essential change here is `ForwardDiff.GradientResult` --> `DiffBase.DiffResult`. I just added the preallocated `GradientConfig`/`HessianConfig` code in this PR because it's a new option that should improve performance.

**Comment by pkofod** [comment]
Sorry, I thought this was from 0.3.0 to 0.3.1. The patch-comment was referring to the .1, but I misunderstood. I really appreciate you took the time to fix it.

**Comment by jrevels** [comment]
No worries, and thanks! Is this good to merge/tag?

**Comment by pkofod** [comment]
> No worries, and thanks! Is this good to merge/tag?

lgtm, I still have to double check if a tag is going to pass the METADATA process. Sorry if Optim is holding up the ecosystem, but I am so busy these days. I'll try to go through the commits to make sure everything is :+1: befores submitting a PR at METADATA.

**Pull request merged**